pkg/controller/storagecluster/storagecluster.go

```diff
@@ -778,6 +779,19 @@ func (c *Controller) updateStorageClusterStatus(
 	cluster *corev1.StorageCluster,
 ) error {
 	toUpdate := cluster.DeepCopy()
+	// Phase and condition list can be updated by different components, e.g. migration handler,
+	// so get the latest status before updating
+	liveCluster := &corev1.StorageCluster{}
+	err := c.client.Get(context.TODO(),
+		types.NamespacedName{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+		}, liveCluster)
+	if err != nil {
+		return err
+	}
+	toUpdate.Status.Phase = liveCluster.Status.Phase
+	toUpdate.Status.Conditions = liveCluster.Status.Conditions
```

Comment:
If we are getting overwriting the phase and conditions from what is in API server, then why are we calling `updateStorageClusterStatus`? Doesn't the caller want to do the same?

Reply:
we call this controller `updateStorageClusterStatus` only when reconcile finished, while other conditions like migration or preflight are updated by the client directly and reflects in the stc object without calling this, otherwise when it hits error before reaching here the conditions will never be updated.
other components can update conditions and write to the object as well in component reconcile loops in the future, but here it should the only place where we iterate all conditions to conclude the phase, and supposed to be called when reconcile finished.

Reply:
major issue to resolve here is conditions are updated by components, so we only overwrite the conditions here (phase should be optional). then underneath in `Driver.UpdateStorageClusterStatus`, px runtime condition will be updated first, then we iterate all conditions to conclude the phase.

Reply:
found a better solution here, check and set conditions all in one place `Driver.UpdateStorageClusterStatus` if possible, including marking migration as completed. I think this makes more sense and easier to maintain. pr updated.